src/box/lua/slab.cc

```diff
@@ -265,6 +275,137 @@ lbox_slab_check(MAYBE_UNUSED struct lua_State *L)
 	return 0;
 }
 
+/*
+ * Decodes and returns the XML document returned by malloc_info() as is.
+ *
+ * This is an internal function that isn't supposed to be called by users,
+ * but it may be useful for developers.
+ *
+ * Returns an empty table if malloc_info() isn't supported by the system.
+ */
+static int
+lbox_malloc_internal_info(struct lua_State *L)
+{
+#if defined(HAVE_MALLOC_INFO)
+	char *buf;
+	size_t buf_size;
+	FILE *fp = open_memstream(&buf, &buf_size);
+	if (fp == NULL) {
+		diag_set(SystemError, "failed to open memory stream");
+		return luaT_error(L);
+	}
+	if (malloc_info(0, fp) != 0) {
+		diag_set(SystemError, "failed to get malloc info");
+		return luaT_error(L);
+	}
+	fclose(fp);
+	lua_pushlstring(L, buf, buf_size);
+	free(buf);
+	return luaT_xml_decode(L);
+#else /* !defined(HAVE_MALLOC_INFO) */
+	lua_newtable(L);
+	return 1;
+#endif /* defined(HAVE_MALLOC_INFO) */
+}
+
+/*
+ * Returns the malloc memory usage information in a table
+ *
+ *   {
+ *     size = <total allocated>,
+ *     used = <actually used>,
+ *   }
+ *
+ * (all numbers are in bytes).
+ *
+ * The information is retrieved with malloc_info(). If it isn't supported by
+ * the system or its format is unknown, {size = 0, used = 0} is returned.
+ */
+static int
+lbox_malloc_info(struct lua_State *L)
+{
+	int version = 0;
+	uint64_t total = 0;
+	uint64_t available = 0;
+	lbox_malloc_internal_info(L);
```

Comment:
Shouldn't it be `pcall`'ed?

Reply:
`box.malloc.internal.info()` may fail because of:
1. `open_memstream`. The only reason I can think of why it may fail is OOM, but in this case tarantool is likely to panic anyway (because we use `xmalloc`).
2. `malloc_info`. According to the manual, this function may fail only on invalid arguments.
3. `luaT_xml_decode`. If `malloc_info` returns an invalid XML document, this will raise an error, but this means that something is seriously broken.

Neither of the three cases should normally happen. Of course, `malloc_info` may be overridden by a third-party allocator, but if it's broken it's better to raise an error instead of silencing this. I added a comment:

```diff
diff --git a/src/box/lua/slab.cc b/src/box/lua/slab.cc
index 0006a59f02d1..63c7f4c78e2c 100644
--- a/src/box/lua/slab.cc
+++ b/src/box/lua/slab.cc
@@ -327,6 +327,12 @@ lbox_malloc_info(struct lua_State *L)
 	int version = 0;
 	uint64_t total = 0;
 	uint64_t available = 0;
+	/*
+	 * box.malloc.internal.info() may raise only if something is seriously
+	 * wrong, for example, we have a bug in the XML parser or malloc_info
+	 * is broken. Either way, it's better to draw the attention by throwing
+	 * an error to the user.
+	 */
 	lbox_malloc_internal_info(L);
 	/*
 	 * The XML document name returned by malloc_info is expected to be
```

Reply:
Better for whom? We already discussed that a user unlikely expect statistics and logging to break a business logic. Your proposal was to report zeros from the public function and report all the details from the internal function. I would follow it consistently, not only for some kinds of an unexpected result.

If it really matters to show the errors from the public function, how about adding them to a field in the result (and keep `size` and `used` as zeros)?

```lua
tarantool> box.malloc.info()
- used: 0
  size: 0
  error: "slab.cc:298: failed to get malloc info"
```

Reply:
Fair enough. Let's call `box.malloc.internal.info()` with `pcall` then. As for adding the `error` field to the stats table, I don't think it's necessary, because a user application collecting stats would silently ignore it anyway.

```diff
diff --git a/src/box/lua/slab.cc b/src/box/lua/slab.cc
index 9d9e5e92e1c5..2e0ba9e680ef 100644
--- a/src/box/lua/slab.cc
+++ b/src/box/lua/slab.cc
@@ -282,6 +282,7 @@ lbox_slab_check(MAYBE_UNUSED struct lua_State *L)
  * but it may be useful for developers.
  *
  * Returns an empty table if malloc_info() isn't supported by the system.
+ * Raises a Lua error if it fails to retrieve or parse malloc_info() output.
  */
 static int
 lbox_malloc_internal_info(struct lua_State *L)
@@ -322,6 +323,8 @@ lbox_malloc_internal_info(struct lua_State *L)
  *
  * The information is retrieved with malloc_info(). If it isn't supported by
  * the system or its format is unknown, {size = 0, used = 0} is returned.
+ *
+ * This function never raises.
  */
 static int
 lbox_malloc_info(struct lua_State *L)
@@ -329,13 +332,9 @@ lbox_malloc_info(struct lua_State *L)
 	int version = 0;
 	uint64_t total = 0;
 	uint64_t available = 0;
-	/*
-	 * box.malloc.internal.info() may raise only if something is seriously
-	 * wrong, for example, we have a bug in the XML parser or malloc_info
-	 * is broken. Either way, it's better to draw the attention by throwing
-	 * an error to the user.
-	 */
-	lbox_malloc_internal_info(L);
+	lua_pushcfunction(L, lbox_malloc_internal_info);
+	if (luaT_call(L, 0, 1) != 0)
+		goto out;
 	/*
 	 * The XML document name returned by malloc_info is expected to be
 	 * "malloc" so the document content should be in malloc[1].
```